kv_store/lww_kvs/kvs_proxy.cpp

```diff
@@ -100,90 +159,637 @@ int main(int argc, char* argv[]) {
+        // serialize request and send
+        string key_req;
+        it->second.SerializeToString(&key_req);
+        zmq_util::send_string(key_req, &requesters[it->first.key_exchange_connect_addr_]);
+
+        // wait for a response from the server and deserialize
+        string key_res = zmq_util::recv_string(&requesters[it->first.key_exchange_connect_addr_]);
+        communication::Key_Response server_res;
+        server_res.ParseFromString(key_res);
+
+        string key;
+        address_t worker_id;
+        address_t gossip_addr;
+        // get the worker address from the response and sent the serialized
+        // data from up above to the worker thread; the reason that we do
+        // this is to let the metadata thread avoid having to receive a
+        // potentially large request body; since the metadata thread is
+        // serial, this could potentially be a bottleneck; the current way
+        // allows the metadata thread to answer lightweight requests only
+        for (int i = 0; i < server_res.tuple_size(); i++) {
+          key = server_res.tuple(i).key();
+          if (it->first.tier_ == "E") {
+            for (int j = 0; j < server_res.tuple(i).address_size(); j++) {
+              worker_id = server_res.tuple(i).address(j).addr();
+              vector<string> v;
+              split(worker_id, ':', v);
+              gossip_addr = worker_node_t(v[0], stoi(v[1]) - SERVER_PORT).distributed_gossip_connect_addr_;
+              communication::Gossip_Tuple* tp = gossip_map[gossip_addr].add_tuple();
+              tp->set_key(key);
+              tp->set_value(key_value_map[key].first);
+              tp->set_timestamp(key_value_map[key].second);
             }
           } else {
-            zmq_util::send_string("succeed status is " + to_string(response.succeed()) + "\n", &user_responder);
+            // we only have one address for memory tier
+            worker_id = server_res.tuple(i).address(0).addr();
+            vector<string> v;
+            split(worker_id, ':', v);
+            gossip_addr = worker_node_t(v[0], stoi(v[1]) - SERVER_PORT).distributed_gossip_connect_addr_;
+            communication::Gossip_Tuple* tp = gossip_map[gossip_addr].add_tuple();
+            tp->set_key(key);
+            tp->set_value(key_value_map[key].first);
+            tp->set_timestamp(key_value_map[key].second);
           }
-        } else {
-          zmq_util::send_string("No servers available.\n", &user_responder);
-        }
-        
-        request.Clear();
-        server_req.Clear();
-        res.Clear();
-        response.Clear();
-      }     
+        }
+      }
+
+      // send as distributed gossips to worker nodes
+      for (auto it = gossip_map.begin(); it != gossip_map.end(); it++) {
+        string data;
+        it->second.SerializeToString(&data);
+        zmq_util::send_string(data, &pushers[it->first]);
+      }
+
+
+
+      // code below is kept just in case we want to change the design...
```

Comment:
I don't think we should need to keep this in the code. It's clunky and hard to read. We can find it in old commits if necessary. :-)